### PR TITLE
[improve][client] Add a way to configure which DNS use

### LINKED ISSUE #21226: Add a configurable DNS server list
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

Currently is not possibile to configure one or more dns servers that will be used by the client to resolve hostnames.

### Solution

In #13390 and 13503 has been added the capability to altering the dns binding address, we can enhance it more allowing DNS server list too

### Alternatives

You can try to set dns server addresses using netty properties like `io.netty.resolver.dns.defaultNameServerFallback` (see DefaultDnsServerAddressStreamProvider) but it doesn't work on jdk 8 nor on Android systems.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by diegosalvi**
This closes #21226

### Motivation

Currently is not possibile to configure one or more dns servers that will be used by the client to resolve hostnames.

### Modifications

Added an optional configuration ClientConfigurationData.dnsServerAddresses to allow to configure custom DNSs. It not provided or empty will default to the existing behaviour.
Handled the new conf in ConnectionPool.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - *Added unit test on ClientBuilderImplTest with configuration of both good or wrong dns servers*
  - *Added integration test on PulsarClientImplTest with a custom configured dns servers list*
  - *Added unit test on ConfigurationDataUtilsTest about the new configuration loading*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [x] `doc` 
- [ ] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/ClientBuilderImplTest.java [hunk 1]
```diff
@@ -101,6 +103,17 @@ public void testClientBuilderWithIllegalLargePort() throws PulsarClientException
         PulsarClient.builder().dnsLookupBind("localhost", 65536).build();
     }
 
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void testClientBuilderWithIllegalDNSServerHostname() throws PulsarClientException {
+        PulsarClient.builder().dnsServerAddresses(
+                Arrays.asList(new InetSocketAddress("1.2.3.4", 53), new InetSocketAddress("localhost",53))).build();
+    }
+
+    public void testClientBuilderWithDNSServerIP() throws PulsarClientException {
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply pulsar-client/src/test/java/org/apache/pulsar/client/impl/ClientBuilderImplTest.java:113]: Added

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/conf/ConfigurationDataUtilsTest.java [hunk 2]
```diff
@@ -61,8 +63,10 @@ public void testLoadClientConfigurationData() {
         config.put("authParams", "testAuthParams");
         config.put("authParamMap", authParamMap);
         config.put("dnsLookupBindAddress", "0.0.0.0");
-        config.put("dnsLookupBindPort", 0);
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply pulsar-client/src/test/java/org/apache/pulsar/client/impl/conf/ConfigurationDataUtilsTest.java:64]: Good catch! It was totally a wrong "cleanup"!

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/ClientBuilderImplTest.java [hunk 3]
```diff
@@ -101,6 +103,18 @@ public void testClientBuilderWithIllegalLargePort() throws PulsarClientException
...
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void testClientBuilderWithIllegalDNSServerHostname() throws PulsarClientException {
+        PulsarClient.builder().dnsServerAddresses(
+                Arrays.asList(new InetSocketAddress("1.2.3.4", 53), new InetSocketAddress("localhost",53))).build();
+    }
+
+    @Test()
+    public void testClientBuilderWithDNSServerIP() throws PulsarClientException {
+        PulsarClient.builder().dnsServerAddresses(
+                Arrays.asList(new InetSocketAddress("1.2.3.4", 53))).build();
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply pulsar-client/src/test/java/org/apache/pulsar/client/impl/ClientBuilderImplTest.java]: Done

**Comment by diegosalvi** [comment]
Rebased on synced fork

**Comment by diegosalvi** [comment]
Last review issue has been addressed

**Comment by diegosalvi** [comment]
Fails org.apache.pulsar.common.nar.NarUnpackerTest.shouldExtractFilesOnceInDifferentProcess
Could it be flaky?

**Comment by lhotari** [comment]
@diegosalvi Could you please provide more context of the use case? The reason why I'm asking is that I don't currently see why the Pulsar client should have a different DNS server configured apart from the rest of the application and the OS where the app is running.

**Comment by lhotari** [comment]
> Fails org.apache.pulsar.common.nar.NarUnpackerTest.shouldExtractFilesOnceInDifferentProcess Could it be flaky?

Yes, it's flaky. reported as #21291

**Comment by diegosalvi** [comment]
@lhotari currently we are using Pulsar in a production environment. In such environment we have need to use different dns for the server machine (configured by standard resolvconf et simillia) and java processes. For many java cases we directly do lookups on dns with different libraries (so we don't use java discovered dnses) but we can't do the same for Pulsar.
To be able to configure which dns use can be useful depending on environments to specialize them

**Comment by diegosalvi** [comment]
@nodece Rebased

**Comment by lhotari** [comment]
> @lhotari currently we are using Pulsar in a production environment. In such environment we have need to use different dns for the server machine (configured by standard resolvconf et simillia) and java processes. For many java cases we directly do lookups on dns with different libraries (so we don't use java discovered dnses) but we can't do the same for Pulsar.
> To be able to configure which dns use can be useful depending on environments to specialize them

@diegosalvi for future documentation purposes, would you be able to share more details and a common reference? Is it a split DNS setup where you have an internal DNS separate from an external DNS ? Would you configure the Pulsar client to use the internal DNS server?

**Comment by diegosalvi** [comment]
@lhotari Yes, something along the line. DNS for Pulsar are "internal" "applicative" ones leaving the host machine with a different set of dns (via resolv.conf etc etc) and other "external" DNS queried directly by java processes when in need to resolve "public" names.

**Comment by diegosalvi** [comment]
Still fails for NarUnpackerTest

**Comment by lhotari** [comment]
> Still fails for NarUnpackerTest

@diegosalvi This flaky NarUnpackerTest has been fixed. You can get the fix to this PR by rebasing on or merging origin/master.

**Comment by diegosalvi** [comment]
Rebased again

**Review by tisonkun — COMMENTED** [review]
I agree with this proposal and since it adds a new public API, I suppose a PIP is required to create, while it should be very straight forward and we can quickly reach a consensus.

@diegosalvi what do you think? You can refer to [this page](https://github.com/apache/pulsar/blob/master/pip/README.md) for sending a PIP.

**Comment by tisonkun** [comment]
Pending for merging as PIP-305 #21352 merged.

**Review by tisonkun — APPROVED** [review]
Thank you!

**Pull request merged**